src/monmove.cpp

```diff
@@ -89,6 +119,39 @@ bool monster::can_move_to( const tripoint &p ) const
         }
     }
 
+    if( has_flag( MF_PATH_AVOID_DANGER ) ) {
+        // Don't enter sharp terrain unless tiny, or attacking
+        if( g->m.has_flag( "SHARP", p ) && !( attitude( &( g->u ) ) == MATT_ATTACK ||
+                                              type->size == MS_TINY || has_flag( MF_FLIES ) ) ) {
+            return false;
+        }
+
+        // Don't enter open pits ever unless tiny, can fly or climb well
+        if( !( type->size == MS_TINY || can_climb ) &&
+            ( target == t_pit || target == t_pit_spiked || target == t_pit_glass ) ) {
+            return false;
+        }
+
+        // Don't enter lava ever
+        if( target == t_lava ) {
+            return false;
+        }
+
+        // Don't enter any dangerous fields
+        if( is_dangerous_fields( g->m.field_at( p ) ) ) {
+            return false;
+        }
+
+        // And don't step on any traps (if we can see)
+        if( has_flag( MF_SEES ) && !g->m.tr_at( p ).is_benign() && g->m.has_floor( p ) ) {
+            return false;
+        }
+
+        if( g->m.has_flag( TFLAG_NO_FLOOR, p ) && !has_flag( MF_FLIES ) ) {
```

Comment:
No clue, honestly, that particular bit of code is copied basically verbatim from the animal logic above it. If you know of a preferred one to use for both rather than two different checks I'm all ears.